## 🎄 Day 1 - Not Quite Lisp
Santa was hoping for a white Christmas, but his weather machine's "snow" function is powered by stars, and he's fresh out! To save Christmas, he needs you to collect fifty stars by December 25th.

Collect stars by helping Santa solve puzzles. Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

Here's an easy puzzle to warm you up.

Santa is trying to deliver presents in a large apartment building, but he can't find the right floor - the directions he got are a little confusing. He starts on the ground floor (floor `0`) and then follows the instructions one character at a time.

An opening parenthesis, `(`, means he should go up one floor, and a closing parenthesis, `)`, means he should go down one floor.

The apartment building is very tall, and the basement is very deep; he will never find the top or bottom floors.

For example:

* `(())` and `()()` both result in floor `0`.
* `(((` and `(()(()(` both result in floor `3`.
* `))(((((` also results in floor `3`.
* `())` and `))(` both result in floor `-1` (the first basement level).
* `)))` and `)())())` both result in floor `-3`.

To **what floor** do the instructions take Santa?

In [3]:
with open("data/2015_puzzle1.txt", "r") as f:
  instr = f.read()

In [5]:
import numpy as np

up = np.count_nonzero(np.array(list(instr)) == "(")
down = np.count_nonzero(np.array(list(instr)) == ")")
up - down

232

Your puzzle answer was `232`. ✅

#### Part Two
Now, given the same instructions, find the position of the first character that causes him to enter the basement (floor `-1`). The first character in the instructions has position `1`, the second character has position `2`, and so on.

For example:

* `)` causes him to enter the basement at character position `1`.
* `()())` causes him to enter the basement at character position `5`.

What is the **position** of the character that causes Santa to first enter the basement?

In [6]:
floor = 0
for n, i in enumerate(list(instr)):
  if i == "(":
    floor += 1
  else:
    floor -= 1
  if floor == -1:
    print(n+1)
    break

1783


Your puzzle answer was `1783`. ✅

## 🎄 Day 2 - I Was Told There Would Be No Math
The elves are running low on wrapping paper, and so they need to submit an order for more. They have a list of the dimensions (length `l`, width `w`, and height `h`) of each present, and only want to order exactly as much as they need.

Fortunately, every present is a box (a perfect right rectangular prism), which makes calculating the required wrapping paper for each gift a little easier: find the surface area of the box, which is `2*l*w + 2*w*h + 2*h*l`. The elves also need a little extra paper for each present: the area of the smallest side.

For example:

* A present with dimensions `2x3x4` requires `2*6 + 2*12 + 2*8 = 52` square feet of wrapping paper plus `6` square feet of slack, for a total of `58` square feet.
* A present with dimensions `1x1x10` requires `2*1 + 2*10 + 2*10 = 42` square feet of wrapping paper plus `1` square foot of slack, for a total of `43` square feet.

All numbers in the elves' list are in feet. How many total **square feet of wrapping paper** should they order?

In [15]:
with open("data/2015_puzzle2.txt", "r") as f:
  sizes = f.readlines()

In [45]:
A_tot = 0
for s in sizes:
  l, w, h = s[:-1].split("x")
  l = int(l); w = int(w); h = int(h)
  A_min = min(min(l*w, w*h), h*l)
  A = 2*l*w + 2*w*h + 2*h*l + A_min
  A_tot += A

In [46]:
print(A_tot)

1598415


Your puzzle answer was `1598415`. ✅

#### Part Two
The elves are also running low on ribbon. Ribbon is all the same width, so they only have to worry about the length they need to order, which they would again like to be exact.

The ribbon required to wrap a present is the shortest distance around its sides, or the smallest perimeter of any one face. Each present also requires a bow made out of ribbon as well; the feet of ribbon required for the perfect bow is equal to the cubic feet of volume of the present. Don't ask how they tie the bow, though; they'll never tell.

For example:

* A present with dimensions `2x3x4` requires `2+2+3+3 = 10` feet of ribbon to wrap the present plus `2*3*4 = 24` feet of ribbon for the bow, for a total of `34` feet.
* A present with dimensions `1x1x10` requires `1+1+1+1 = 4` feet of ribbon to wrap the present plus `1*1*10 = 10` feet of ribbon for the bow, for a total of `14` feet.

How many total **feet of ribbon** should they order?

In [47]:
l_tot = 0
for s in sizes:
  l, w, h = s[:-1].split("x")
  l = int(l); w = int(w); h = int(h)
  wrap = 2 * min(min(l+w, w+h), h+l)
  bow = l*w*h
  l_tot += wrap + bow

In [48]:
print(l_tot)

3812909


Your puzzle answer was `3812909`. ✅

## 🎄 Day 3 - Perfectly Spherical Houses in a Vacuum
Santa is delivering presents to an infinite two-dimensional grid of houses.

He begins by delivering a present to the house at his starting location, and then an elf at the North Pole calls him via radio and tells him where to move next. Moves are always exactly one house to the north (`^`), south (`v`), east (`>`), or west (`<`). After each move, he delivers another present to the house at his new location.

However, the elf back at the north pole has had a little too much eggnog, and so his directions are a little off, and Santa ends up visiting some houses more than once. How many houses receive **at least one present**?

For example:

* `>` delivers presents to `2` houses: one at the starting location, and one to the east.
* `^>v<` delivers presents to `4` houses in a square, including twice to the house at his starting/ending location.
* `^v^v^v^v^v` delivers a bunch of presents to some very lucky children at only `2` houses.

In [49]:
with open("data/2015_puzzle3.txt", "r") as f:
  dirs = f.read()

In [52]:
coords = [[0, 0]]
for d in list(dirs):
  new_coord = coords[-1].copy()
  if d == "^":
    new_coord[1] += 1
  elif d == "v":
    new_coord[1] -= 1
  elif d == ">":
    new_coord[0] += 1
  elif d == "<":
    new_coord[0] -= 1
  coords.append(new_coord)

In [57]:
len(np.unique(coords, axis=0))

2572

Your puzzle answer was `2572`. ✅

#### Part Two
The next year, to speed up the process, Santa creates a robot version of himself, **Robo-Santa**, to deliver presents with him.

Santa and Robo-Santa start at the same location (delivering two presents to the same starting house), then take turns moving based on instructions from the elf, who is eggnoggedly reading from the same script as the previous year.

This year, how many houses receive **at least one present**?

For example:

* `^v` delivers presents to `3` houses, because Santa goes north, and then Robo-Santa goes south.
* `^>v<` now delivers presents to `3` houses, and Santa and Robo-Santa end up back where they started.
* `^v^v^v^v^v` now delivers presents to `11` houses, with Santa going one direction and Robo-Santa going the other.

In [58]:
santa_coords = [[0, 0]]
robo_coords = [[0, 0]]

for i, d in enumerate(list(dirs)):
  if (i % 2) == 0:
    new_coord = santa_coords[-1].copy()
  else:
    new_coord = robo_coords[-1].copy()

  if d == "^":
    new_coord[1] += 1
  elif d == "v":
    new_coord[1] -= 1
  elif d == ">":
    new_coord[0] += 1
  elif d == "<":
    new_coord[0] -= 1

  if (i % 2) == 0:
    santa_coords.append(new_coord)
  else:
    robo_coords.append(new_coord)

In [61]:
len(np.unique(np.concatenate([santa_coords, robo_coords]), axis=0))

2631

Your puzzle answer was `2631`. ✅